### ISSUE TITLE
Data::Printer "bugfix". Please read and comment!

### ISSUE THREAD

**Issue by garu** [garu/Data-Printer#16]
Data::Printer has a problem with return values. By default, it tries to be smart for the user, so if you do:

```perl
   use DDP; p $data;
```

it will dump the data to STDERR. But if you do:

```perl
   use DDP; my $ret = p $data;
```

it will return the dump instead of printing it. This also let's you do stuff like:

```perl
   use DDP; print p $data;
```

in case you want the dump to go to STDOUT instead of STDERR. Of course, you can also achieve this
by doing something like:

```perl
  use DDP output => 'stdout'; p $data;
```

But I digress.

The problem is that, to detect whether the user wants us to return something or not, we rely
on Perl's "wantarray" function. This is all nice and well, except if you look at wantarray's documentation
you'll see a minor catch:

>   "wantarray()"'s result is unspecified in the top level of a
>   file, in a "BEGIN", "UNITCHECK", "CHECK", "INIT" or "END"
>   block, or in a "DESTROY" method.

This means that if you're debugging stuff in BEGIN, DESTROY, etc, you might do something like:

```perl
   use DDP; p $data;
```

and get absolutely *nothing*. The reverse would also happen, where you could do:

```perl
   use DDP; my $str = p $data;
```

and see it printed on STDERR instead of returned. Of course, since you'll see the data on the screen,
this is less annoying than the former.

I have received some reports that `p($data)` "wasn't working" meaning it wasn't printing anything on the screen,
and after a lot of debugging I think `wantarray`'s odd behavior might be the cause - specially since the user was within a BEGIN block.

Another huge annoyance where Data::Printer doesn't DWIM is when it's the very last statement of your function.
Let's say you have this:

```perl
  sub do_something {
     ...
  }

  my $data = do_something();
```

and you're debugging the `do_something()` function, so you want to call `p()` to see your data before you
continue coding. If you did something like:

```perl
  my $data = do_something();
  use DDP; p $data;
```

then you'd be just fine. The problem lies when you try to mangle inside do_something(). And why
shouldn't you? I mean, you're debugging, right? So you do something like:

```perl
  sub do_something {
     ...
     use DDP; p $struct;
  }

  my $data = do_something;
```

and you get *nothing* on the output. WTF?

#### Did you spot what happended?

If you check the docs, you'll see _mst_ warned about this months ago (but who wants
to look at the docs? It's a tool to print data structures. You want it to work, dammit!)

The problem here is that you want something from do_something(), and since you called
p() as the last statement, you basically told perl that do_something() should return
whatever it gets from p(), meaning now p() is **NOT** in void context, so it returns
the dump instead of printing it on the screen.

If you had made sure p() wasn't the last statement in the function ( p $struct; 1; ) or
if you had called do_something() in void context, or if you had set DDP with the 'pass' return
value (which triggers the pass-through behavior in which p() will always print and
return the value of the actual variable, not the dump string), then everything would have
worked just fine.

*Except you didn't, and now you're debugging the debugging tool. Not cool.*

At first I thought it wasn't that big a deal, that it was a very specific debugging scenario
and the user would likely known what it was doing. So as long as it was documented,
everything would be fine. Turns out it's not that specific at all, and
the very last thing you want when debugging code is to be bitten by your debugging tool.

I clearly had made a poor design decision, and now that Data::Printer is being used all over (thanks, guys!)
I feel like you should help me out with making the _right_ decision this time, if you want/care.

## What should I do? What should Data::Printer do?

Leon Timmermans already helped by pointing out a way to query caller() and see if we're inside the
BEGIN, INIT, etc, blocks, but this doesn't solve the latter issue, which can be even more annoying.

As I see it (and I'm **VERY** open to other ideas) I have a few ways to "fix" this:

=> Leave it as it is, and add a bigger note on such behavior. I don't like this one, otherwise
I wouldn't be having this argument. But I'm adding it just for thorougness' sake.

=> make p() pass-through by default. This will make p() print the dump to STDERR *always*
unless you do something about it, where something is:

```perl
   use DDP return_value => 'dump'; # go back to the (now, so far) default behavior
   use DDP output => \$var;        # append the dump to $var
   use DDP output => 'stdout';     # use STDOUT instead of STDERR
```

If we change the default behavior to that, it means that if you ever did something like:

```perl
   use DDP; print p $data;
```

or

```perl
   use DDP; warn p $data;
```

You'll get $data dumped to STDERR and will print whatever is in `$data` back to STDOUT (via the `print()` call).

It also means that if you wanted the dump string back, you'd have to change `return_value` back to "dump" or
point the `output` property to a scalar reference, like in the first two examples above.

Do people really use the return values? I think they can be quite useful for something like printing to a
debugging webpage, or maybe a log file.

Which brings me to solution #2:

=> Change p()'s default return_value behavior to 'pass' just like in #1, but also exporting a d() function
that would be like p() except always returning the string instead of printing, just like Data::Dumper does.

The reason I'm not in love with this idea is because I don't want to pollute the caller's namespace with
more functions - one should be enough I guess.

We could of course add an alias for d() like we do for p(), or even make the exporting optional. But if the
user has to adjust his/hers "`use DDP`" parameters, then why can't he/she simply make the changes above?

Which brings me to potential solution #3:

=> Change p()'s default return_value behavior to 'pass' just like in #1, but create a new package called,
say, DDX (heck, we could even use DDP itself so it wouldn't be just an alias do Data::Printer), that
would export 2 functions to the user's namespace (and not just one, as Data::Printer would):

   p() - always prints stuff
   d() - always returns the string

This could even be enough to fix another annoyance to more advanced Data::Printer users: prototypes.

   p() - always prints stuff, uses prototypes
   np() - always prints stuff, no prototypes
   d()  - always returns the string, uses prototypes

Does it look sane-ish to you? Not really? Alternatives?

It could be the case of creating solution #4 in which Data::Printer's p() won't change its default
behavior, but DDP's p() will. But I'm not sure I want to maintain this sort of thing, so I'm more inclined to #2 or #3.

Of course, I'd have to make sure all other customization options are in sync with whatever is chosen
as solution.

**Comment by marcelgruenauer** [comment]
+1. I've run into this behaviour as well, and while I've shrugged it off as an edge case, fixing it is good.

**Comment by oalders** [comment]
+1

**Comment by hoelzro** [comment]
:+1:

**Comment by jayallen** [comment]
+1! Nice work @garu

**Comment by oalders** [comment]
@garu is there a target date for releasing this?  I see that #57 references a Feb 2015 release.  :)

**Comment by hernan604** [comment]
@oalders Rumors says the release will go live this weekend \o/

**Comment by genehack** [comment]
The new behavior broke code I had; -1 for the backward incompatible change.

**Comment by jayallen** [comment]
@genehack Out of curiosity, how did it break precisely?  What behavior change did you see and what's the context of the code?

**Comment by genehack** [comment]
@jayallen - 

Here's the code that was broken: https://github.com/genehack/HiD/blob/master/lib/HiD/App/Command/config.pm#L46

(The fix was simply adding the `return_value => 'dump'` bit to the `use Data::Printer` line.)

Here's what a failing test looked like: 
http://www.cpantesters.org/cpan/report/dba5c99c-ef79-11e4-8d26-63e2dfbfc7aa

It was not at all obvious, until I really sat down and walked through stuff carefully, and noticed that the tester was using a dev version of Data::Printer, that I was able to track down it was this change. 

The thing I don't like about the solution is, I use DDP when developing this code, frequently throwing 'p $foo' statements into code. I haven't tested, but I suspect mixing `use DDP` and `use DDP return_value => 'dump'` statements in the same code base is going to eventually cause pain.

Once the changes are fully released, I guess I require that version and switch to the `np()` function call; right now, with the mix of old and new behaviors circulating in the wild, what I've done feels like the best of a variety of bad options.

**Comment by garu** [comment]
@genehack I am really sorry you ran into trouble due to my poor design decision in the early days of Data::Printer. `return value => 'dump'` might be indeed the best of a variety of bad options. Sadly, having this RFC open since early 2012 and commented on by several great developers, I think it's time we get this over with and push the new default behaviour. Unless of course you have a better idea, in which case I'd be more than willing to "stop the presses" :)

**Comment by genehack** [comment]
@garu pushing the new default behavior out will allow me to just depend on the new version, and then the issue is fixed from my point of view -- so, please, full speed ahead. 8^)

**Comment by garu** [comment]
@exodist @shadowcat-mst @melo @doherty @bessarabov @aduitsis @frioux @marcelgruenauer @oalders @hoelzro @jayallen @hernan604 @genehack (and everyone else who even bothered to *look* at this issue):

It is done! After 3+ years, I finally took the courage to break stuff and release this backwards-incompatible (but much needed) change as 0.36. There's plenty more changes to come in the future too, including fixes to some outstanding issues that have long been postponed because of it. Stay tuned!

Also, **thank you so much** for your time and patience, and for helping me fix the root problem in the least intrusive and violent way possible. You rock!!!

**Comment by Elenadamon** [comment]
You have to Continue to export p() as you do now with the exact same behavior, do not make a backwards-incompatible change. and  also export pout() and perr() that always print to stderr and stdout regardless of context. and you can also take reference from the [epson printer error ](http://www.printererrorrepair.com/blog/how-to-fix-epson-printer-error-turn-power-off-and-then-on-again/) and you will find the relevant solution.

**Comment by printersupportnumber001** [comment]
In that case, one should not opt the option of always print. For some time, it would work for you. But later it will become more than worst. So, as per my perception, you should go for the newly updated device. Why waste time on unnecessary fixes. Or visit **[Printer Support Service](http://technicalsupportforprinter.com)**, perhaps it would resolve your issues.

**Comment by techcustomersupport** [comment]
For the bug fix to take effect, both firmware 2.5 and profiles 6.0 need to be installed. Use [Connect PS4 Controller](https://macsupportnumber.com/blog/how-to-connect-ps4-controller-to-mac/) service and get the best solution regarding how to connect PS4 controller to mac.
